torchtitan/parallelisms/parallelize_llama.py

```diff
@@ -215,27 +244,39 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
         assert dp_mesh.mesh_dim_names == ("dp",), dp_mesh.mesh_dim_names
         # TODO: Expose `reduce_dtype` as a config option.
         mp_policy = MixedPrecisionPolicy(
-            param_dtype=torch.bfloat16, reduce_dtype=torch.float32
+            # TODO(whc) need to fix PP + FSDP-mixed-precision
+            # tracer for PP assumes f32 and is caught off guard when runtime FSDP interacts using bf16 inputs
+            # param_dtype=torch.bfloat16, reduce_dtype=torch.float32
+            param_dtype=torch.float32,
```

Comment:
we shouldn't by default change this, this would make the cases where FSDP or FSDP + TP use fp32 instead of bf16

Reply:
I wonder if supporting bf16 should be a criteria for landing. I would imagine that training with FSDP + PP in fp32 is not really viable efficiency-wise (at least for larger jobs).

Reply:
I think we should fix this before landing the PP change.  I think there was a possible way to fix this in the tracer, but lost track of it, will dig it up